### PR TITLE
Cast on host instead of IPU when using `precision=16` 

### LINKED ISSUE #13828: Cast happens after copy in mixed precision training on IPU
## 🐛 Bug

During mixed precision training on IPU, the cast happens after the copy. When working in half precision on IPU it is preferred to cast to cast to half precision on the host.

### To Reproduce

<details>
<summary>
repro.py
</summary>

```python
import os
import torch
import argparse
import poptorch
import pytorch_lightning as pl
import torch.nn.functional as F
from torchvision import datasets, models
from torchvision.transforms import ToTensor
from pytorch_lightning.strategies.ipu import IPUStrategy


class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = models.resnet18(pretrained=False)

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.model(x)
        loss = F.cross_entropy(y_hat, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)


def main():
    batch_size = 32
    # data
    mnist = datasets.FakeData(
        size=100 * batch_size,
        image_size=(3, 28, 28),
        num_classes=10,
        transform=ToTensor(),
    )
    opts = poptorch.Options()
    train_loader = poptorch.DataLoader(
        options=opts,
        dataset=mnist,
        batch_size=batch_size,
        drop_last=True,
        mode=poptorch.DataLoaderMode.Async,
    )
    # model
    model = Model()
    # training
    trainer = pl.Trainer(
        accelerator="ipu",
        precision=args.precision,
        max_epochs=1,
        strategy=IPUStrategy(training_opts=opts),
    )
    trainer.fit(model, train_loader)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--precision", choices=[16, 32], default=16, type=int)
    args = parser.parse_args()
    os.environ["POPLAR_ENGINE_OPTIONS"] =f'{{"autoReport.all":"true","autoReport.directory":"profile_fp{args.precision}"}}'
    main()
```

</details>

```bash
python repro.py --precision 16
python repro.py --precision 32
```

Look at the profiles in PopVision Graph Analyser and see that the copy is FP32 followed by a cast to FP16 on the IPU.

### Expected behavior

In the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#precision), the video tells us that the weights remain FP32 and the data is cast to the selected precision. When training with FP16 data on IPU, the cast should occur before copying the data to device. The screenshots below show that the data is FP32 during the copy when `precision=16`.

### Environment

* Packages:
        - numpy:             1.23.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cpu
        - pytorch-lightning: 1.6.5
        - tqdm:              4.64.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #124-Ubuntu SMP Thu Apr 14 19:46:19 UTC 2022

### Additional context

FP16 - left, FP32 - right

In this image you can see that the total amount of data received by the IPU for this batch is 301184B in both cases. If we divide by the size of the batch, it can be seen that the datatype in both copies is FP32. `301184 / (32*3*28*28) = 4B = 32b`:
image

cc @carmocca @justusschock @awaelchli @akihironitta @rohitgr7

### PR DESCRIPTION
**Pull request by HMellor**
## What does this PR do?

Removes the cast to half from the forward pass and adds it to the host side pre-processing.

In overloading `batch_to_device` it also prevents `data.to(device)`, which is not currently necessary for IPU.

Fixes #13828

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:265]: since this wasn't overridden before, was it sending it to the IPU device before using the super class method?

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:265]: also, now at what point does it send the batch to the IPU device?

Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:265]: > since this wasn't overridden before, was it sending it to the IPU device before using the super class method?

This used to happen in the forward pass in the `LightningIPUModule`, which was too late. Because by this time the data was already on the IPU.

> also, now at what point does it send the batch to the IPU device?

This is handled by PopTorch via [StreamCopies](https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/user-guide.html#streamcopy) and shouldn't be done explicitly using `data.to(device)`

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:265]: so I think we can keep the `LightningIPUModule` and override `batch_to_device` to just return the batch and not do anything. This way it will be cast to fp16 first and then moved to the device inside the `forward` method.

also can you add a test?

Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:265]: It can't be done in `forward` because `forward` is traced and therefore all operations in forward are compiled and uploaded to the IPU.

Would a test that checks the type of the input to `LightningIPUModule.forward` be good?

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:44]: we can remove this? since it doesn't do anything now.

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:259]: let's make this a local function inside `batch_to_device`?

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:262]: can you add a comment here?

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:265]: I think yet, or if there's any existing test, you can add some assertions there.

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:265]: > The data.to(device) that is currently called doesn't actually do anything

that means `self.device`doesn't point to the actual IPU device?

File: src/pytorch_lightning/strategies/ipu.py
Comment (awaelchli) [reply src/pytorch_lightning/strategies/ipu.py:44]: If the logic here is really not needed, I'm also in favor (removal with deprecation)

File: src/pytorch_lightning/strategies/ipu.py
Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:265]: > that means `self.device` doesn't point to the actual IPU device?

I'm not sure what is supposed to have `device` as an attribute (it's not an attribute of the `IPUStrategy`), but the `device` argument of `batch_to_device` is `None` when running my reproducer.

File: src/pytorch_lightning/strategies/ipu.py
Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:44]: I have removed `LightningIPUModule`, but I'm not sure what you're deprecation strategy is so please let me know if there's anything else I need to do

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:123]: IPUs can only trace the `forward` method? not the `training_step/validation_step`, etc...?

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:44]: to deprecate the class, keep it and do:
```python
class LightningIPUModule(_LightningModuleWrapperBase):
    def __init__(self, ... same old args):
        rank_zero_deprecation('`LightningIPUModule` has been deprecated in v1.8.0 and will be removed in v2.0.0')
        ... same old code
```
and a test in `tests/tests_pytorch/deprecated_api/test_remove_2-0.py`.

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:259]: if that's the case, we need to do a follow-up for users to let them know that some of the hooks won't work with IPUs, like `transfer_batch_to_device`, ...

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py]: `self.device` is a general attribute that has the value of current device. Gives the ability to users to create tensors anytime and move them to device even if they are not coming from the dataloaders. That means this is not possible when using IPUs, right?

one can't do
```py
def training_step(...):
    some_tensor = torch.randn(...).to(self.device)  # this won't move the newly created tensor to IPU?
```

Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py]: > That means this is not possible when using IPUs, right?

That is correct

File: src/pytorch_lightning/strategies/ipu.py
Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:123]: The `forward` method of the `model` argument passed to `poptorch.{training/inference}Model` is currently traced. Before this PR, the `LightningIPUModule` was being passed as the `model` argument, so its `forward` was being traced. This was why the cast was happening on the IPU instead of the host.

File: src/pytorch_lightning/strategies/ipu.py
Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py:283]: Is this something you could do? I don't know the codebase well enough to know all the implications.

File: src/pytorch_lightning/strategies/ipu.py
Comment (awaelchli) [inline src/pytorch_lightning/strategies/ipu.py:49]:
```suggestion
        rank_zero_deprecation("`LightningIPUModule` is deprecated in v1.8 and will be removed in v1.10")
```

File: tests/tests_pytorch/deprecated_api/test_remove_2-0.py
Comment (awaelchli) [inline tests/tests_pytorch/deprecated_api/test_remove_2-0.py:57]:
```suggestion
    with pytest.deprecated_call(match=r"is deprecated in v1.8 and will be removed in v1.10."):
```

File: tests/tests_pytorch/accelerators/test_ipu.py
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_ipu.py:43]: These lines would fail without the fix in the strategy, correct?

**Review by awaelchli — APPROVED** [review]
Thank you for working on this!

File: src/pytorch_lightning/strategies/ipu.py
Comment (awaelchli) [reply src/pytorch_lightning/strategies/ipu.py:49]: If we merge it as bugfix before 1.7 is release, we should decrement the numbers here. Waiting for reviews and decision.

Comment (carmocca) [reply src/pytorch_lightning/strategies/ipu.py:49]: I propose we just give this a 1 version deprecation.

And I agree this can be considered a bug-fix. So let's assume it will be deprecated in `v1.7.x` and removed in `v1.8`.

Meaning we update these messages, cherry-pick into a bug-fix release after merging this PR, and then open a new PR removing the code from master.

File: src/pytorch_lightning/CHANGELOG.md
Comment (carmocca) [inline src/pytorch_lightning/CHANGELOG.md:116]: Since we will be considering this change a bug-fix, please move this into the "Fixed" section

File: tests/tests_pytorch/deprecated_api/test_remove_2-0.py
Comment (carmocca) [reply tests/tests_pytorch/deprecated_api/test_remove_2-0.py:57]: And also move it to the correct file

File: src/pytorch_lightning/strategies/ipu.py
Comment (HMellor) [reply src/pytorch_lightning/strategies/ipu.py]: I've updated this to deprecate in 1.7 and remove in 1.8

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/accelerators/test_ipu.py
Comment (HMellor) [reply tests/tests_pytorch/accelerators/test_ipu.py:43]: It passes now, but I haven't verified that it would have failed before. We shouldn't merge until I have

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:276]:
```suggestion
            batch = apply_to_collection(batch, Tensor, function=Tensor.half)
```
if we need to move all kinds of tensors for IPUs

File: tests/tests_pytorch/deprecated_api/test_remove_1-8.py
Comment (rohitgr7) [inline tests/tests_pytorch/deprecated_api/test_remove_1-8.py:1013]:
```suggestion
    with pytest.deprecated_call(match=r"has been deprecated in v1.7.0 and will be removed in v1.8."):
```

File: src/pytorch_lightning/strategies/ipu.py
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:49]:
```suggestion
        rank_zero_deprecation("`LightningIPUModule` has been deprecated in v1.7.0 and will be removed in v1.8.0")
```

**Review by rohitgr7 — APPROVED** [review]
LGTM

File: tests/tests_pytorch/accelerators/test_ipu.py
Comment (HMellor) [reply tests/tests_pytorch/accelerators/test_ipu.py]: Since `batch_to_device` is the last thing that happens before the tensors are _supposed_ to be on the device, the updated test should be valid. It fails before the PR and passes after.

File: src/pytorch_lightning/strategies/ipu.py
Comment (carmocca) [reply src/pytorch_lightning/strategies/ipu.py]: Is this the change that caused https://github.com/Lightning-AI/lightning/pull/13983?

Comment (justusschock) [reply src/pytorch_lightning/strategies/ipu.py]: that's part of it, but before e.g. DoubleTensors, BfloatTensors etc. wouldn't have been cast as well.

**Pull request merged**
